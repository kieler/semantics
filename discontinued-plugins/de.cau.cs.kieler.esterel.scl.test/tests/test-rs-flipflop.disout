module RSFlipFlop:
function calc_or(boolean, boolean) : boolean;
function calc_not(boolean) : boolean;
procedure dump()(boolean, boolean);
input SET : boolean;
input RESET : boolean;
output ITERATE;
output ITERATE12;
output Q : boolean;
output NQ : boolean;

signal O_OR1 := false : boolean,
       O_OR2 := false : boolean,
       O_NOT1 := false : boolean,
       O_NOT2 := false : boolean,
       O_PORT12 := false : boolean in
    loop
      abort
        loop
          pause
        end loop
      when
        case SET
        case O_NOT2
      end abort;
      var rv : boolean in
        rv := calc_or(?SET, ?O_NOT2);
        emit O_OR1(rv)
      end var
    end loop
  ||
    loop
      abort
        loop
          pause
        end loop
      when
        case O_PORT12
        case RESET
      end abort;
      var rv : boolean in
        rv := calc_or(?O_PORT12, ?RESET);
        emit O_OR2(rv)
      end var
    end loop
  ||
    loop
      abort
        loop
          pause
        end loop
      when O_OR1;
      var rv : boolean in
        rv := calc_not(?O_OR1);
        emit O_NOT1(rv)
      end var
    end loop
  ||
    loop
      abort
        loop
          pause
        end loop
      when O_OR2;
      var rv : boolean in
        rv := calc_not(?O_OR2);
        emit O_NOT2(rv)
      end var
    end loop
  ||
    var old : boolean,
        value : boolean in
      old := false;
      value := false;
      emit O_PORT12(false);
      abort
        loop
          pause
        end loop
      when O_NOT1;
      old := ?O_NOT1;
      emit ITERATE12;
      pause;
      emit O_PORT12(old);
      loop
        abort
          loop
            pause
          end loop
        when O_NOT1;
        value := ?O_NOT1;
        call dump()(old, value);
        if (value <> old) {
          emit ITERATE12;
          pause;
          emit O_PORT12(value);
          old := value
        }
      end loop
    end var
  ||
    every O_NOT1 do
      emit NQ(?O_NOT1)
    end every
  ||
    every O_NOT2 do
      emit Q(?O_NOT2)
    end every
  ||
    loop
      abort
        loop
          pause
        end loop
      when ITERATE12;
      emit ITERATE
    end loop
end signal

end module
