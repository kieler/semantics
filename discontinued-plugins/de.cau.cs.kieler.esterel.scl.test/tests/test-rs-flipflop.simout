   0 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
   1 ITERATE=1 ITERATE12=1 Q=1 (1) NQ=1 (0) 
   2 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
   3 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
   4 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
   5 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
   6 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
   7 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (0) 
   8 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
   9 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  10 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  11 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  12 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (0) 
  13 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  14 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  15 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  16 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (0) 
  17 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  18 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  19 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  20 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  21 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  22 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  23 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  24 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  25 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  26 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  27 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  28 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  29 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  30 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  31 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  32 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  33 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  34 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  35 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  36 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  37 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  38 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  39 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  40 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (1) 
  41 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (1) 
  42 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (0) 
  43 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  44 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  45 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  46 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  47 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  48 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  49 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  50 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  51 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (0) 
  52 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  53 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  54 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  55 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  56 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  57 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  58 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  59 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  60 ITERATE=1 ITERATE12=1 Q=0 NQ=1 (0) 
  61 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  62 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  63 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  64 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  65 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  66 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  67 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  68 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  69 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  70 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  71 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  72 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  73 ITERATE=1 ITERATE12=1 Q=0 NQ=1 (0) 
  74 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  75 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  76 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  77 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  78 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  79 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  80 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  81 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  82 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  83 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  84 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  85 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  86 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  87 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  88 ITERATE=0 ITERATE12=0 Q=0 NQ=0 
  89 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  90 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  91 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (1) 
  92 ITERATE=1 ITERATE12=1 Q=0 NQ=1 (0) 
  93 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (0) 
  94 ITERATE=0 ITERATE12=0 Q=1 (1) NQ=1 (0) 
  95 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (0) 
  96 ITERATE=1 ITERATE12=1 Q=1 (0) NQ=1 (1) 
  97 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
  98 ITERATE=0 ITERATE12=0 Q=0 NQ=1 (1) 
  99 ITERATE=0 ITERATE12=0 Q=1 (0) NQ=1 (1) 
