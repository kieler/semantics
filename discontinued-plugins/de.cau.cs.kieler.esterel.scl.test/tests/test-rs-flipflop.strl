%
% Implementation eines RS FlipFlops auf Basis von
% or und not. Kommunikation zwischen den Modulen
% erfolgt über "Port-Automaten"
%
%

module RSFlipFlop:

input SET   : boolean;	% Eingangs Werte
input RESET : boolean;  

output ITERATE;
output ITERATE12;

output Q	 : boolean; % Ausgänge
output NQ    : boolean;


signal	O_OR1 := false : boolean,
		O_OR2 := false  : boolean, 
		O_NOT1:= false  : boolean, 
		O_NOT2:= false : boolean,
		O_PORT12 := false  : boolean
in
	[
		run OR
		[
			signal SET/a, O_NOT2/b, 
				   O_OR1/out
		];
	||
		run OR
		[
			signal RESET/b, O_PORT12/a, 
				   O_OR2/out
		];
	||
		run NOT
		[
			signal O_OR1/v, 
			       O_NOT1/out
		];
	||
		run NOT
		[
			signal O_OR2/v, 
			       O_NOT2/out
		];
	||
		run BOOLPORT
		[
			signal	O_NOT1/IN,
					O_PORT12/OUT, ITERATE12/ITERATE
		];
	||
		% Stub: Ausgaben auf Modulausgaben mappen
		every O_NOT1 do
			emit NQ ( ?O_NOT1 );
		end
	||
		every O_NOT2 do
			emit Q ( ?O_NOT2 );
		end
	||
		% Stub: Change weiterleiten
		loop
			await ITERATE12;
			
			emit ITERATE
		end
	]

end 

.
module OR:

input 	a   : boolean;	% Eingangswerte
input   b   : boolean;

output  out : boolean;

function calc_or (boolean, boolean) :boolean;

loop

	% Sensitive auf Eingaben warten (a oder b)
	await
		case a
		case b
	end;
	
	% Ausgabe senden
	var rv : boolean in
		rv:=calc_or (?a,?b);
		emit out (rv);
	end;
%	emit out ( ?a or ?b );

end

.
module NOT:

input 	v    : boolean;	% Eingangswert
output 	out  : boolean;	% Ausgabe

function calc_not ( boolean ) : boolean;

loop

	% Sensitive auf Input warten
	await v;
	
	var rv : boolean in
		
		rv:=calc_not ( ?v );
		emit out (rv);
	end
	
	% Ausführen
%	emit out ( not ?v );

end

.
%
% Port für den Datentyp Boolean
%
% Ausgang wird nur gesetzt, wenn Eingang
% anders als der alte Wert
%

module BOOLPORT:

input	IN   : boolean;
output  OUT  : boolean;
output  ITERATE;

procedure dump ()(boolean,boolean);

var	old   := false : boolean, 
	value := false : boolean
in

	%% Initiale Phase
	emit OUT (false);
	
	await IN;
	old:=?IN;
	emit ITERATE;
	pause;

	emit OUT (old);
	
	%% Hauptschleife		
		loop
			
			% Auf Wert warten und entgegennehmen
			await IN;	
			value:=?IN;
			
			call dump ()(old,value);
			
			% Senden, wenn Werte unterschiedlich
			if ( value <> old ) then

				% Takt warten
				emit ITERATE;
				pause;
			
				emit OUT (value);
				old:=value;
			end;

		end 
	
end

.
