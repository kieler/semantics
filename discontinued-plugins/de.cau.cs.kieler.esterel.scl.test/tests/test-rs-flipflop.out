module RSFlipFlop:
input SET : boolean;
input RESET : boolean;
output ITERATE;
output ITERATE12;
output Q : boolean;
output NQ : boolean;

signal O_OR1 := false : boolean,
       O_OR2 := false : boolean,
       O_NOT1 := false : boolean,
       O_NOT2 := false : boolean,
       O_PORT12 := false : boolean in
    run OR [
      signal SET / a,
             O_NOT2 / b,
             O_OR1 / out ]
  ||
    run OR [
      signal RESET / b,
             O_PORT12 / a,
             O_OR2 / out ]
  ||
    run NOT [
      signal O_OR1 / v,
             O_NOT1 / out ]
  ||
    run NOT [
      signal O_OR2 / v,
             O_NOT2 / out ]
  ||
    run BOOLPORT [
      signal O_NOT1 / IN,
             O_PORT12 / OUT,
             ITERATE12 / ITERATE ]
  ||
    every O_NOT1 do
      emit NQ(?O_NOT1)
    end every
  ||
    every O_NOT2 do
      emit Q(?O_NOT2)
    end every
  ||
    loop
      await ITERATE12;
      emit ITERATE
    end loop
end signal

end module

module OR:
function calc_or(boolean, boolean) : boolean;
input a : boolean;
input b : boolean;
output out : boolean;

loop
  await 
    case a
    case b
  end await;
  var rv : boolean in
    rv := calc_or(?a, ?b);
    emit out(rv)
  end var
end loop

end module

module NOT:
function calc_not(boolean) : boolean;
input v : boolean;
output out : boolean;

loop
  await v;
  var rv : boolean in
    rv := calc_not(?v);
    emit out(rv)
  end var
end loop

end module

module BOOLPORT:
procedure dump()(boolean, boolean);
input IN : boolean;
output OUT : boolean;
output ITERATE;

var old := false : boolean,
    value := false : boolean in
  emit OUT(false);
  await IN;
  old := ?IN;
  emit ITERATE;
  pause;
  emit OUT(old);
  loop
    await IN;
    value := ?IN;
    call dump()(old, value);
    if value <> old then
      emit ITERATE;
      pause;
      emit OUT(value);
      old := value
    end if
  end loop
end var

end module
